File: src/Java/algorithms/treeandgraph/SumRootToLeafNumbers.java
package algorithms.treeandgraph; 

import java.util.*; 
import lib.TreeNode; 

public class SumRootToLeafNumbers {
    /* Function: Find the sum of all paths in the tree presented as number. 
     * Solution: previous path nodes sum * 10 add this node is the value of 
     *     this path to this node. Notice only when left and right all null, 
     *     it can count as a finished path. 
     */ 
    
    public int sumNumbers(TreeNode root) {
        return sumHelper(root, 0); 
    }
    
    private int sumHelper(TreeNode root, int pre) {
        if (root == null) {
            return 0; 
        }
        
        int cum = pre * 10 + root.val; 

        if (root.left == null && root.right == null) {
            return cum; 
        }
        
        return sumHelper(root.left, cum) + sumHelper(root.right, cum); 
    }
}

File: src/CPP/src/leetcode/SumRootToLeafNumbers.cpp
/*
 * [Source] https://leetcode.com/problems/sum-root-to-leaf-numbers/
 * [Difficulty]: Medium
 * [Tag]: Tree
 * [Tag]: Depth-first Search
 */

#include <iostream>

using namespace std;

// [Solution]: Previous * 10 + current is the number.
// [Corner Case]: If this node doesn't has left and right child, it is done.
class Solution {
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/treeandgraph/SumRootToLeafNumbers.java
 */

/* Java solution
public class Solution {
    public int sumNumbers(TreeNode root) {
		if(root==null) return 0;
		return sumNumbers(root, 0);
    }
	
	public int sumNumbers(TreeNode root, int sum){
		if(root==null) return 0;
		int s=sum*10+root.val;
		if(root.left==null && root.right==null) return s;
		return sumNumbers(root.left, s)+sumNumbers(root.right, s);
    }
}
 */

int main() {
    Solution sol;

    return 0;
}
